core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -3103,25 +3116,46 @@ private int updatedDeliveryCount(DeliveryCountOps ops) {
             };
         }
 
-        private void archive(String newMemberId) {
-            state = RecordState.ARCHIVED;
-            memberId = newMemberId;
+        // Visible for testing.
+        void archive(String newMemberId) {
+            stateTransitionLock.writeLock().lock();
+            try {
+                if (rollbackState != null) {
+                    isMarkedArchived = true;
+                }
+                state = RecordState.ARCHIVED;
+                memberId = newMemberId;
+            } finally {
+                stateTransitionLock.writeLock().unlock();
+            }
         }
 
-        private InFlightState startStateTransition(RecordState newState, DeliveryCountOps ops, int maxDeliveryCount, String newMemberId) {
-            rollbackState = new InFlightState(state, deliveryCount, memberId, acquisitionLockTimeoutTask);
-            return tryUpdateState(newState, ops, maxDeliveryCount, newMemberId);
+        // Visible for testing
+        InFlightState startStateTransition(RecordState newState, DeliveryCountOps ops, int maxDeliveryCount, String newMemberId) {
+            stateTransitionLock.writeLock().lock();
+            try {
+                rollbackState = new InFlightState(state, deliveryCount, memberId, acquisitionLockTimeoutTask);
+                return tryUpdateState(newState, ops, maxDeliveryCount, newMemberId);
```

Comment:
Should it happen as below:
```suggestion
                InFlightState currentState = new InFlightState(state, deliveryCount, memberId, acquisitionLockTimeoutTask);
                InflightState updatedState = tryUpdateState(newState, ops, maxDeliveryCount, newMemberId);
                if (updatedState != null) {
                      rollbackState = currentState;
                }
                return updatedState;
```